Document agentless Feature Flags delivery for Node.js (#38231)
DataDog/documentation: content/en/feature_flags/server/_index.md

- link: "/feature_flags/implementation_patterns/serverless/"
  tag: "Documentation"
  text: "Serverless environments and Feature Flags"
- link: "/feature_flags/concepts/configuration_sources/"
  tag: "Concept"
  text: "Server SDK Configuration Sources"
---

## Overview

> Datadog Feature Flags for server-side applications allow you to remotely control feature availability, run experiments, and roll out new functionality with confidence. Server-side SDKs integrate with the Datadog tracer for your language, receive Universal Flag Configuration (UFC), and evaluate flags locally.

Author: ufc

Author: Addressed in [4df3f4876b2359d8a2de96b0344c555787ec8da6](https://github.com/DataDog/documentation/commit/4df3f4876b2359d8a2de96b0344c555787ec8da6): Replaced UFC with flag configuration.

> Datadog Feature Flags is built on the [OpenFeature standard](https://openfeature.dev/docs/reference/intro/), an open source, vendor-neutral specification for feature flag APIs. If you're new to OpenFeature concepts like providers, evaluation context, and hooks, see the [OpenFeature concepts documentation](https://openfeature.dev/docs/category/concepts).

Author: Not applying this suggestion. The repository Vale rule explicitly requires `open source` and flags `open-source`; the cited sibling page is itself producing that warning. Keeping the current wording.

## Configuration delivery

Agentless [configuration delivery][8] is the default in server SDK versions that support it. The SDK fetches flag configuration directly from the Datadog-managed CDN over HTTPS, then evaluates flags locally. A Datadog Agent is not required for flag configuration.

The default source does not activate Feature Flags traffic for every tracer installation. Agentless polling begins only when application code initializes or accesses the Datadog OpenFeature provider. Explicitly selecting `remote_config` activates the Feature Flags Remote Configuration subscription. Requests through either source contribute to server Feature Flags billing.

| SDK | Minimum agentless version |
|---|---|
| Node.js `dd-trace` v5 | 5.116.0 |
| Node.js `dd-trace` v6 | 6.5.0 |

<div class="alert alert-warning">The initial Node.js agentless releases support configuration delivery and local flag evaluation only. They do not support evaluation metrics, exposure logging, or experimentation use cases.</div>

> Agentless delivery is available for Node.js in the versions listed. Other server SDKs use Agent Remote Configuration until their language pages list an agentless minimum version.

Reviewer: This phrasing will be more future-proofed as additional SDKs start supporting agentless:
```suggestion
Agentless delivery is available for the SDKs and versions listed. Other server SDKs use Agent Remote Configuration.
```

  {{< image-card href="/feature_flags/server/ruby/" src="integrations_logos/ruby.png" alt="Ruby" >}}
{{< /card-grid >}}

For serverless runtimes, see [Serverless Environments][5] for no-Agent setup, version requirements, and initial telemetry limitations.

## Prerequisites

Default Datadog-managed server SDK setups require:

- **APM tracing** [enabled in your application][4]
- The language-specific tracer and OpenFeature provider versions listed on the SDK page
- A Datadog [API key][2]

Source-specific requirements are:

| Source | Requirements |
|---|---|
| `agentless` (default where supported) | Configure `DD_API_KEY`, `DD_ENV`, and optionally `DD_SITE` in the application process. No Agent is required for flag configuration. |
| `remote_config` | Datadog Agent 7.55 or later with Remote Configuration enabled, the API key configured on the Agent, and Remote Configuration enabled for your organization in [{{< ui >}}Organization Settings{{< /ui >}}][3]. |

## Agentless configuration

> For Node.js 5.116.0 or 6.5.0, configure the application process:

Author: Addressed in [4df3f4876b2359d8a2de96b0344c555787ec8da6](https://github.com/DataDog/documentation/commit/4df3f4876b2359d8a2de96b0344c555787ec8da6): Clarified that 5.116.0 and 6.5.0 are dd-trace versions for Node.js.

{{< code-block lang="bash" >}}
# Required for direct configuration delivery
DD_API_KEY=<DATADOG_API_KEY>
DD_ENV=<YOUR_ENVIRONMENT>
{{< /code-block >}}

> No Feature Flags enablement or source setting is required. If your organization is not on the default `datadoghq.com` site, also set `DD_SITE`. See the Node.js page for dependency versions and language-specific initialization. Initializing or accessing the provider starts CDN polling; tracer installation and initialization alone do not.

Reviewer: The link here depends on adding a reference entry at the end of the file (comment below):
```suggestion
No Feature Flags enablement or source setting is required. If your organization is not on the default {{< region-param key="dd_site" code="true" >}} site, also set `DD_SITE`. See [Node.js Feature Flags][] for dependency versions and language-specific initialization. Initializing or accessing the provider starts CDN polling; tracer installation and initialization alone do not.
```

Author: Thanks for showing how to do that dynamically.

> ## Agent remote configuration

Author: Addressed in [4df3f4876b2359d8a2de96b0344c555787ec8da6](https://github.com/DataDog/documentation/commit/4df3f4876b2359d8a2de96b0344c555787ec8da6): Capitalized the Remote Configuration product name.

For Node.js, set the source explicitly to retain Agent-managed delivery:

{{< code-block lang="bash" >}}
DD_FEATURE_FLAGS_CONFIGURATION_SOURCE=remote_config
{{< /code-block >}}

Remote Configuration is enabled by default in Agent 7.47.0 and later. If your Agent has Remote Configuration disabled, re-enable it by setting `DD_REMOTE_CONFIGURATION_ENABLED=true` or adding `remote_configuration.enabled: true` to your `datadog.yaml`.

See the [Remote Configuration documentation][1] for detailed setup instructions across deployment environments.

> Existing Node.js customers who set `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true` remain on Remote Configuration during a migration window. The setting is deprecated. See [Migrate from the legacy provider setting][7] to remain on Remote Configuration explicitly or move to agentless delivery.

Reviewer: ```suggestion
Existing Node.js implementations with `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true` remain on Remote Configuration during a migration window. The setting is deprecated. See [Migrate from the legacy provider setting][7] to remain on Remote Configuration explicitly or move to agentless delivery.
```

> ### Remote configuration polling interval

Author: Addressed in [4df3f4876b2359d8a2de96b0344c555787ec8da6](https://github.com/DataDog/documentation/commit/4df3f4876b2359d8a2de96b0344c555787ec8da6): Capitalized the Remote Configuration product name.

DD_ENV=<YOUR_ENVIRONMENT>
DD_VERSION=<YOUR_APP_VERSION>

# Optional: Disable Feature Flags and both delivery paths
# DD_FEATURE_FLAGS_ENABLED=false

# Optional: Enable flag evaluation metrics
# See "Set Up Server-Side Flag Evaluation Metrics" documentation
{{< /code-block >}}

<div class="alert alert-info">In the Node.js versions listed above, <code>DD_FEATURE_FLAGS_ENABLED</code> defaults to <code>true</code>, so you do not need to set it. Setting it to <code>false</code> disables the provider, CDN polling, and the Feature Flags Remote Configuration subscription. Other server SDKs continue to use the activation settings documented on their language pages.</div>

> For SDKs and delivery modes that support it, see <a href="/feature_flags/guide/server_flag_evaluation_metrics/">Set Up Server-Side Flag Evaluation Metrics</a> to enable the experimental <code>feature_flag.evaluations</code> metric. The initial Node.js agentless releases do not support evaluation metrics, exposure logging, or experimentation use cases. See <a href="/feature_flags/concepts/flag_graphs/">Feature Flag Graphs</a> for more information on available graphing.

Reviewer: Identifying the feature as experimental here sets up a maintenance burden, as we'll have to revise this if the feature becomes no longer experimental in the future. The target page has a warning banner about this, so that helps ensure readers are aware:
```suggestion
For SDKs and delivery modes that support it, see <a href="/feature_flags/guide/server_flag_evaluation_metrics/">Set Up Server-Side Flag Evaluation Metrics</a> to enable the <code>feature_flag.evaluations</code> metric. The initial Node.js agentless releases do not support evaluation metrics, exposure logging, or experimentation use cases. See <a href="/feature_flags/concepts/flag_graphs/">Feature Flag Graphs</a> for more information on available graphing.
```

[3]: https://app.datadoghq.com/organization-settings/remote-config
[4]: /tracing/guide/#tutorials-enabling-tracing
[5]: /feature_flags/implementation_patterns/serverless/

> [6]: /feature_flags/concepts/configuration_sources/

Author: Addressed in [dd6638fbc8b52d5c66c75fc9873cc3f33aac8643](https://github.com/DataDog/documentation/commit/dd6638fbc8b52d5c66c75fc9873cc3f33aac8643): removed the unused server overview reference definition.

> [8]: /feature_flags/concepts/configuration_sources/

Reviewer: Add reference for link added above:
```suggestion
[8]: /feature_flags/concepts/configuration_sources/
[9]: /feature_flags/server/nodejs/
```
